pkg/logger/json.go

```diff
@@ -0,0 +1,274 @@
+/*
+Copyright 2022 CodeNotary, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logger
+
+import (
+	"bytes"
+	"encoding"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	// DefaultTimeFormat is the time format to use for JSON output
+	DefaultTimeFormat = "2006-01-02T15:04:05.000000Z07:00"
+
+	// errInvalidTypeMsg message implies an arg cannot be serialized to json
+	errInvalidTypeMsg = "cannot serialize arg(s) to json"
+
+	// defaultCallerOffset is the stack frame offset in the call stack for the caller
+	defaultCallerOffset = 4
+)
+
+var (
+	// defaultOutput is used as the default log output.
+	defaultOutput io.Writer = os.Stderr
+)
+
+// NewJSONFileLogger returns a json logger which outputs to file
+func NewJSONFileLogger(name string, file string) (logger Logger, out *os.File, err error) {
+	if out, err = setup(file); err != nil {
+		return nil, nil, err
+	}
+	return &JsonLogger{
+		name:       name,
+		timeFormat: DefaultTimeFormat,
+		timeFnc:    time.Now,
+		mutex:      new(sync.Mutex),
+		writer:     newWriter(out),
+		level:      int32(logLevelFromEnvironment()),
+	}, out, nil
+}
+
+// NewJSONLogger returns a json logger.
+func NewJSONLogger(opts *Options) *JsonLogger {
```

Comment:
opts could even be a value of type Options

Reply:
I did not quite understand...

Reply:
You mean, pass by value than reference?